Add support eDMA for S32K344

Dat-NguyenDuy opened the pull request:
This extends NXP current eDMA shim driver for supporting eDMA v3 on S32K344

Dat-NguyenDuy committed:
drivers: dma_mcux_edma: add support for multiple DMAMuxes

This adds support for platforms that have multiple DMAMUXes
per DMA instance

Signed-off-by: Dat Nguyen Duy <dat.nguyenduy@nxp.com>

Dat-NguyenDuy committed:
drivers: dma_mcux_edma: add support for edma version 3

Add new dt binding for edma v3 that inherits whole dt
properties from current version. One more property is
added for SoCs that don't have separate error interrupt
id, use same id with channel interrupt

Signed-off-by: Dat Nguyen Duy <dat.nguyenduy@nxp.com>

Dat-NguyenDuy committed:
drivers: dma_mcux_edma: improve interrupt handling

The current implementation iterates over all channels
even if only several channels share the same irq. This
introduces one more dt property to describe an offset
between two channels share the same interrupt id.

Beside that, the error interrupt must be put as last
element of "interrupt" dt property.

Signed-off-by: Dat Nguyen Duy <dat.nguyenduy@nxp.com>

Dat-NguyenDuy committed:
drivers: dma_mcux_edma: add support dma driver for s32k344

On S32K344, the offset in memory map between each channel
is 0x4000 for most channels, but there is specific case is
between channel 11 and 12 which is 0x1D4000 instead. As a
consequence, 32 channels are divided to two parts: one
starts from channel 0 -> 11. The other is from channel 128
to 145. The channel gap is from 12 -> 127.

For user and data structures in shim driver, the channel's
value comes from 0 --> 31. Above constraint will be counted
when interact with the mcux sdk

Beside that, the DMAMUX register in this platform is very
specific, not in identical with DMAMUX channel, so shim
driver is updated to cover this case

Signed-off-by: Dat Nguyen Duy <dat.nguyenduy@nxp.com>

Dat-NguyenDuy committed:
tests: drivers: dma: enable dma test for mr_canhubk3

Add devicetree overlay for enabling dma test for mr_canhubk3 board

Signed-off-by: Dat Nguyen Duy <dat.nguyenduy@nxp.com>

Dat-NguyenDuy committed:
boards: arm: mr_canhubk3: enable dma support

Enable dma support for mr_canhubk3 board

Signed-off-by: Dat Nguyen Duy <dat.nguyenduy@nxp.com>

Dat-NguyenDuy committed:
drivers: dma_mcux_edma: correct valid values for dma channel/slot

The dma-channels devicetree value - 1 = maximum valid channel
The dma-requests devicetree value - 1 = maximum valid request

Signed-off-by: Dat Nguyen Duy <dat.nguyenduy@nxp.com>

mmahadevan108 commented:
@Dat-NguyenDuy, please take a look at the CI failures.

danieldegrasse commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -519,54 +629,93 @@ static int dma_mcux_edma_init(const struct device *dev)
-		IRQ_CONFIG(n, 16, dma_mcux_edma_error_irq_handler);   \
-								      \
-		LOG_DBG("install irq done");			      \
+#ifdef CONFIG_DMA_MCUX_SINGLE_IRQ_PER_CHANNEL
```
It seems like this macro will install an interrupt handler per each channel of the EDMA. Existing NXP SOCs using the legacy EDMA also have a per channel interrupt, but have a combined error interrupt. Would it be possible to enable per channel interrupts for all EDMA instances, and only enable the error interrupt servicing in the channel interrupt when this Kconfig was set?

Dat-NguyenDuy replied:
Good idea, it's possible, i just re-pushed the update

Dat-NguyenDuy replied:
@danieldegrasse could you pls revisit if this one looks fine to you

hakehuang commented on drivers/serial/uart_mcux_lpuart.c:
```diff
@@ -1140,7 +1140,7 @@ static const struct uart_driver_api mcux_lpuart_driver_api = {
 			DEVICE_DT_GET(DT_INST_DMAS_CTLR_BY_NAME(id, tx)),		       \
 		.dma_channel =								       \
-			DT_INST_DMAS_CELL_BY_NAME(id, tx, mux),				       \
+			DT_INST_DMAS_CELL_BY_NAME(id, tx, channel),			       \
```
why cahnge this from mux to channel?

Dat-NguyenDuy replied:
I reverted to keep the current cells

hakehuang commented on dts/bindings/dma/nxp,mcux-edma.yaml:
```diff
@@ -33,22 +38,11 @@ properties:
-#  The 1st cell specifies the DMAMUX(0 for DMAMUX0 and 1 for DMAMUX1).
-#    Specific request source can only be multiplexed by specific channels
-#    group called DMAMUX.
+#  The 1st cell specifies the DMA channel (not DMAMUX channel) which will be used
```
this is not going to work in our legacy platfrom there are two DMA mux with the same channell numbers. I would suggest that you can define a new v3 dt yaml file, and includes both of them.

Dat-NguyenDuy replied:
If i understand correctly, the current bindings for _dma-cells_ was copied from Linux's implementation. But the implementation in Zephyr is difference as mentioned in the commit msg.

Could you pls give me the name of SoC that will not work with this update? look like current boards in upstream are not be impacted

hakehuang replied:
> If i understand correctly, the current bindings for _dma-cells_ was copied from Linux's implementation. But the implementation in Zephyr is difference as mentioned in the commit msg.

it is referenced from linxu dma dts, but we need change for the real case. the impacted platforms are rt11xx seriese, which has 2 edma mux.

Dat-NguyenDuy replied:
@hakehuang is this still a concern for you?

Dat-NguyenDuy committed:
dts: bindings: correct information in mcux edma bindings

With the current implementation, the 1st cell is not DMAMUX id
as mentioned in the bindings (0 for DMAMUX0 and 1 for DMAMUX1).

Moreover, the referenced Linux bindings is obsoleted, it was
migrated to use yaml syntax

Signed-off-by: Dat Nguyen Duy <dat.nguyenduy@nxp.com>

Dat-NguyenDuy commented on boards/arm/mr_canhubk3/doc/index.rst:
```diff
@@ -56,6 +56,7 @@ ADC SAR       on-chip     adc
 LPSPI         on-chip     spi
 WDT           FS26 SBC    watchdog
 EMAC          on-chip     ethernet
+EMDA          on-chip     dma
```
Thanks, updated

Dat-NguyenDuy commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -102,28 +110,71 @@ struct dma_mcux_edma_data {
+#define DEV_DMAMUX_BASE(dev, idx) ((DMAMUX_Type *)DEV_CFG(dev)->dmamux_base[idx])
+#define DEV_DMAMUX_IDX(dev, ch)	(ch / DEV_CFG(dev)->channels_per_mux)
+
+#ifdef CONFIG_SOC_SERIES_S32K3_M7
```
This is SoC dependency, i'm considering it's worth to add a dt properties for it

manuargue replied:
I'd suggest to add a binding property to remove the SoC dependency in the driver, e.g.
```c
#define DEV_DMAMUX_CHANNEL(dev, ch) ((ch % DEV_CFG(dev)->channels_per_mux) ^ (DEV_CFG(dev)->mux_channel_mask))
```
and use a default of `0` which would be the logical way to access the DMA `CHCFGn` registers. The SoCs that have a different arrangement can change the `mux-channel-mask` (feel free to suggest a better name) as needed, like in the case of K344 to `3`.

Dat-NguyenDuy replied:
I've updated

Dat-NguyenDuy commented on dts/bindings/dma/nxp,mcux-edma.yaml:
```diff
@@ -33,22 +38,11 @@ properties:
 
 dma-cells:
-  - mux
+  - channel
```
I reverted to keep the current cells

manuargue reviewed (commented):
@Dat-NguyenDuy please have a look to the failed builds. Some may be unrelated to these changes but caught on this twister run.

Dat-NguyenDuy commented:
> @Dat-NguyenDuy please have a look to the failed builds. Some may be unrelated to these changes but caught on this twister run.

tests build failure for imxrt because of this issue: https://github.com/zephyrproject-rtos/zephyr/pull/61310#issuecomment-1698233681

for k3, this PR should be merged first https://github.com/nxp-mcuxpresso/mcux-sdk/pull/132

manuargue commented on drivers/dma/Kconfig.mcux_edma:
```diff
@@ -6,13 +6,28 @@
+	default y
+	depends on DT_HAS_NXP_MCUX_EDMA_ENABLED
+	help
+	  Enable legacy EDMA for MCUX series SoCs.
```
I'm curious about the reason to name this "legacy" and not something like EDMA V2 or V1. If later there's a new EDMA version V4 let's say, would that turn V3 into legacy? I'd expect legacy is a piece of sw/hw that was superseded but somehow we need to keep maintaining. I'm not familiar with the platforms supported by MCUX HAL, but is expected that the new platforms will not use anymore this "legacy" EDMA version?

Dat-NguyenDuy replied:
What i understand is that the current EDMA is second version of the DMA IP, the EDMA in S32K344 is third version of the DMA IP. Not sure what is better, should either i kept the current version in upstream as it is (edma) and use dmav3 for the S32K344 or something like edmav2 and edmav3

manuargue replied:
if the existing is V2 and the one you're adding is V3, I'd name it V2 and V3 respectively. But I don't have a strong opinion about it. I'll let other comments if somebody have a strong request to change it.

manuargue replied:
@mmahadevan108 @dleach02 do you have any preference on this? otherwise if you are okay with proposed naming here we can close this comment

mmahadevan108 replied:
@manuargue , I agree.

Dat-NguyenDuy replied:
okay, i've updated

manuargue commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -24,13 +24,21 @@
-	DMAMUX_Type *dmamux_base;
+	DMAMUX_Type **dmamux_base;
+#ifdef CONFIG_DMA_MCUX_EDMA_V3
+	uint32_t channel_gap[2];
```
does it make sense to support the channel gaps only for EDMA V3? it doesn't seem to be tied to the version of the EDMA but to a design of the SoC

Dat-NguyenDuy replied:
Just because currently in upstream, i don't see SoCs use the old version of EDMA has the gap

Dat-NguyenDuy replied:
I made it common for both

Dat-NguyenDuy commented on tests/drivers/dma/scatter_gather/boards/mr_canhubk3.overlay:
```diff
@@ -0,0 +1,15 @@
+
+ / {
+	aliases {
+		dma0 = &edma0;
```
yep, could be. I will update

Dat-NguyenDuy replied:
i've updated

Dat-NguyenDuy commented on dts/bindings/dma/nxp,mcux-edma.yaml:
```diff
@@ -10,6 +10,11 @@ include: dma-controller.yaml
     required: true
+    description: |
+      Specifies base physical address(s) and size of the eDMA registers.
+      The 1st region is eDMA control register's address and size.
```
I've updated

Dat-NguyenDuy commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -510,63 +618,102 @@ static int dma_mcux_edma_init(const struct device *dev)
+		{								\
+			ARG_UNUSED(dev);					\
+										\
+			IRQ_CONFIG(n, 0, dma_mcux_edma_irq_handler);		\
```
It's a bit more complicated than that. Some Socs have each irq for each channel, some have several (2) channels share one irq. I've updated to cover both

Dat-NguyenDuy commented:
@hakehuang I put some changes to address some comments, could you pls help to revisit this PR

manuargue commented:
can you please address the remaining failed builds? https://github.com/zephyrproject-rtos/zephyr/pull/61311/checks?check_run_id=16612602788 they seem related to these changes

Dat-NguyenDuy commented:
> can you please address the remaining failed builds? https://github.com/zephyrproject-rtos/zephyr/pull/61311/checks?check_run_id=16612602788 they seem related to these changes

Ah, forgot to mentioned, the patch for DMA was just integrated with in this PR: https://github.com/zephyrproject-rtos/hal_nxp/pull/270. I'll rebase my PR in hal_nxp to pick-up the latest version, the build failures should be fixed

hakehuang commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -24,17 +24,37 @@
 	void (*irq_config_func)(const struct device *dev);
 };
 
+#if defined FSL_FEATURE_EDMA_HAS_ERROR_IRQ && FSL_FEATURE_EDMA_HAS_ERROR_IRQ
```
can this put into the soc PLATFORM_HAS_ERROR_IRQ? put feature dependences in driver  is not that good.

Dat-NguyenDuy replied:
But how about the offset between channels share the same interrupt, we need to use dt? I also see this kind of macros (`FSL_FEATURE`) are using in various mcux shim drivers

hakehuang replied:
use dts is a good selection, the legacy code with FSL_FEATURE is histric burden.

Dat-NguyenDuy replied:
Is it okay if we treat this as an improvement as the Code Freeze is coming? the dt can be used, but i need to add new properties, then update the driver and dt for all mcux SoC that are affected

Dat-NguyenDuy replied:
@hakehuang is this ok for you?

mmahadevan108 replied:
@Dat-NguyenDuy . Hake is suggesting replacing the `FSL_FEATURE` with something you define in `soc.h`. His suggestion is to use `PLATFORM_HAS_ERROR_IRQ`. I feel that should be similar and an easy change.

Dat-NguyenDuy replied:
I've updated, not just that minor. @hakehuang can you pls revisit

hakehuang replied:
we want you to put things in soc level, not in driver config, and what's worse is that you define a marco that skip our code, why you wlways change others code clogic and not consider to make your code fit in existing code?

Dat-NguyenDuy replied:
Re-pushed as discussed in this https://github.com/zephyrproject-rtos/zephyr/pull/61311#discussion_r1333835313

hakehuang commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -24,17 +24,37 @@
 struct dma_mcux_edma_config {
 	DMA_Type *base;
-	DMAMUX_Type *dmamux_base;
+	DMAMUX_Type **dmamux_base;
```
for rt11xx and before although we have multiply dma mux, but they belongs to different domain, mean they can not be access at the same time. but for S32K, one DMA can use multiply dmamux, this is different.

Dat-NguyenDuy replied:
Yep, i understand your point. The dmamux is declared here is the one that can route DMA sources for the controller, not all dmamuxs available in the SoC. Perhaps the `reg-names` cause confusion, i've updated

hakehuang commented on drivers/dma/dma_mcux_edma.c:
```diff
@@ -510,63 +620,89 @@ static int dma_mcux_edma_init(const struct device *dev)
+	};									\
+	static const struct dma_mcux_edma_config dma_config_##n = {		\
+		.base = (DMA_Type *)DT_INST_REG_ADDR_BY_NAME(n, edma),		\
+		.dmamux_base =	&dmamux_base_##n[0],				\
```
looks like this caculation is not correct.

hakehuang commented on dts/arm/nxp/nxp_rt11xx.dtsi:
```diff
@@ -964,6 +965,7 @@
 			nxp,a_on;
 			reg = <0x40c14000 0x4000>,
 				<0x40c18000 0x4000>;
+			reg-names = "edma", "dmamux0";
```
for rt11xx series, dmamux0 is for M7 domain, and dmamux1 is for LPST domain(m4). using dmamux0 is not correct

Dat-NguyenDuy replied:
ok, so i will consider to update this

Dat-NguyenDuy replied:
I've updated, the dmamux is used is the one that can route DMA sources.

manuargue commented:
@Dat-NguyenDuy seems like a rebase didn't go well :) Removed reviewers that were not automatically added by the CI.

Dat-NguyenDuy commented:
> @Dat-NguyenDuy seems like your rebase didn't go well :) (tried to) remove reviewers that were not automatically added by the CI.

right, was my mistake

hakehuang commented on drivers/dma/Kconfig.mcux_edma:
```diff
@@ -33,4 +49,18 @@ config DMA_MCUX_USE_DTCM_FOR_DMA_DESCRIPTORS
 	  When this option is activated, the descriptors for DMA transfer are
 	  located in the DTCM (Data Tightly Coupled Memory).
 
+config DMA_MCUX_HAS_ERROR_IRQ
```
put to soc kconfig, not here

manuargue replied:
this Kconfig is indicating if the error interrupt is implemented, and is part of the DMA hw IP. So it should be in this Kconfig file. The SoC should select this or not depending on the implementation.

hakehuang replied:
if in this case, put them in dts would be better.

Dat-NguyenDuy replied:
okay, i've pushed to update

manuargue reviewed (dismissed):
It looks good to me, thanks @Dat-NguyenDuy for the improvements to the DMA driver that came along the way :)

hakehuang commented on dts/bindings/dma/nxp,mcux-edma-v2.yaml:
```diff
@@ -0,0 +1,83 @@
+      The last element indicates this interrupt is shared for all channels,
+      0 is starting channel and offset 1 between each.
+
+  no-error-irq:
```
if this is IP related, I suppose it is for EDMA_V3, if this is true, I would not recommend adding the no-error-irq here, but add to v3

Dat-NguyenDuy replied:
i've updated

hakehuang commented on dts/bindings/dma/nxp,mcux-edma-v2.yaml:
```diff
@@ -0,0 +1,83 @@
+      Above property provides range for 17 interrupt ids.
+      16 first elements indicate each interrupt is for only one channel.
+      The last element indicates this interrupt is shared for all channels,
+      0 is starting channel and offset 1 between each.
```
please consider revise this implementation

hakehuang commented:
I am trigger CI for all NXP related EDMA boards, will feadback once ready

Dat-NguyenDuy commented:
> I am trigger CI for all NXP related EDMA boards, will feadback once ready

Thanks

hakehuang commented:
> I am trigger CI for all NXP related EDMA boards, will feadback once ready

regression test done. no issues found.

Dat-NguyenDuy commented:
Having minor update `#ifdef DMA_MCUX_HAS_CHANNEL_GAP` --> `#if DMA_MCUX_HAS_CHANNEL_GAP`

hakehuang reviewed (dismissed):
Test PASS on NXP regression for RT series

mmahadevan108 commented:
@Dat-NguyenDuy , thank you for the PR. Overall this looks great. My request would be not leave the current EDMA driver Kconfigs and device tree binding names unchanged to maintain backwards compatibility.

mmahadevan108 reviewed (dismissed):
Request to leave the EDMA Kconfigs and device tree binding names unchanged.

manuargue commented:
> Request to leave the EDMA Kconfigs and device tree binding names unchanged.

@mmahadevan108 I'm a bit puzzled, I had the impression we agreed naming existing binding v2 and the new one v3 to follow the corresponding DMA versions (https://github.com/zephyrproject-rtos/zephyr/pull/61311#discussion_r1331484150). We should add this to release notes to make sure the changes are visible to users.

Dat-NguyenDuy commented:
> Request to leave the EDMA Kconfigs and device tree binding names unchanged.

I re-pushed to leave current EDMA Kconfig and dts binding

mmahadevan108 commented:
Please update `west.yml`

Dat-NguyenDuy commented:
> Please update `west.yml`
I've updated

hakehuang commented:
regression test is done. no issue found in rt platforms with this change
